Fix issue with paging when going to the next page after initial arrive by search

VillePihlava opened the pull request:
### Summary

When doing an arriveby search, the search window can include itineraries that are not included because of the arriveby time limit. With the current implementation in the following example an arrive by search for 16:00 does not include the itinerary in red in the initial search or the next page. By setting the next page earliest departure time based on the latest departure time from the results, the next page contains the missing result.

![image](https://github.com/user-attachments/assets/3fa433bd-9464-4093-8a04-d865332134ec)


### Issue

An alternative solution to https://github.com/opentripplanner/OpenTripPlanner/issues/6607. Instead of adding an overlap this PR fixes an edge case when doing an arrive by search and searching for the next page.

### Unit tests

I changed existing unit tests to take into account the new change

### Documentation

Comments added

miklcct commented:
Why is it 1 minute rather than 1 second?

VillePihlava commented:
> Why is it 1 minute rather than 1 second?

As there was a precedent of using 1 minute, I used it in this PR too:

https://github.com/opentripplanner/OpenTripPlanner/blob/9a0fc4d189742b492d1fcf4abbed5de94277dde6/application/src/main/java/org/opentripplanner/model/plan/paging/cursor/PageCursorFactory.java#L125-L128

t2gran commented on application/src/main/java/org/opentripplanner/service/paging/PagingService.java:
```diff
@@ -183,6 +185,20 @@ private PageCursorFactory mapIntoPageCursorFactory(@Nullable PageType currentPag
       latestArrivalTime,
       searchWindowUsed
     );
+    // If the first search is an arrive by search, add the latest itinerary's start time to allow setting cursor information correctly
+    // for the page cursor of the next page.
+    if (
+      currentPageType == null &&
+      PageCursorFactory.resolvePageTypeForTheFirstSearch(itinerariesSortOrder) == PREVIOUS_PAGE
+    ) {
+      if (itineraries.size() > 0) {
+        factory = factory.withFirstSearchLatestItineraryDeparture(
+          itineraries.get(0).startTimeAsInstant()
+        );
+      } else {
+        factory = factory.withFirstSearchLatestItineraryDeparture(earliestDepartureTime);
+      }
+    }
```
This block should be replaced with a private method - this is a very minor case, and now it totally dominates the logic in this method.

VillePihlava replied:
I created a `handleFirstSearch` method

t2gran commented on application/src/test/java/org/opentripplanner/service/paging/PS2_ManyParetoOptimalItinerariesTest.java:
```diff
@@ -134,11 +134,13 @@ void testPagingArriveByAndNext() {
 
     var nextCursor = subject.nextPageCursor();
 
+    // The earliest departure time is 12:26 here instead of 12:30,
+    // because it is determined from the latest itinerary departure from the results.
```
```suggestion
    // The earliest departure time is 12:26 here, because it is determined from the 
    // latest itinerary departure from the results.
```

VillePihlava replied:
I changed this comment

VillePihlava commented on application/src/test/java/org/opentripplanner/service/paging/PS3_FewItinerariesOnSearchWindowLimitTest.java:
```diff
@@ -91,7 +92,7 @@ static List<Arguments> testCases() {
   @ParameterizedTest
```
I uncommented the three test cases that were commented

optionsome commented on application/src/main/java/org/opentripplanner/service/paging/PagingService.java:
```diff
@@ -184,12 +186,33 @@ private PageCursorFactory mapIntoPageCursorFactory(@Nullable PageType currentPag
+    ) {
+      if (itineraries.size() > 0) {
+        factory = factory.withFirstSearchLatestItineraryDeparture(
+          itineraries.get(0).startTimeAsInstant()
```
So these are always sorted by the departure time, not by arrival time (i.e. if an itinerary starts later but arrives earlier than another one, is that first or second in the list)?

optionsome commented on application/src/main/java/org/opentripplanner/service/paging/PagingService.java:
```diff
@@ -184,12 +186,33 @@ private PageCursorFactory mapIntoPageCursorFactory(@Nullable PageType currentPag
+          itineraries.get(0).startTimeAsInstant()
+        );
+      } else {
+        factory = factory.withFirstSearchLatestItineraryDeparture(earliestDepartureTime);
```
Is it possible with this case or any other case, that this pr actually adjusts the search window  to start later for the next page than what would have been without the changes from this pr?

VillePihlava replied:
There might be a case in which this happens. If the latest itinerary departure happens to be on the last possible minute of the search window (or maybe on second 59:59 in an hour long search window), then the added minute on the end can create a 59 second gap (assuming this uses seconds to determine inclusivity/exclusivity).

https://github.com/opentripplanner/OpenTripPlanner/blob/bb6bc6538239c55b26be149c41a1db92d7b12f75/application/src/main/java/org/opentripplanner/model/plan/paging/cursor/PageCursorFactory.java#L125-L129

The added time should be a minute/second based on how the inclusivity/exclusivity works. This should be discussed in the next dev meeting. Assuming that there are some `Duration` equality checks under the hood, the value should probably be a second.

optionsome commented on application/src/test/java/org/opentripplanner/service/paging/PS2_ManyParetoOptimalItinerariesTest.java:
```diff
@@ -152,7 +154,7 @@ void testPagingArriveByAndNext() {
         PageType.PREVIOUS_PAGE,
         SortOrder.STREET_AND_DEPARTURE_TIME,
-        "12:00",
+        "11:56",
```
Why are both next and previous page cursors affected by this change?

VillePihlava replied:
What's happening here is that the page cursors are tested after one next page cursor has already been requested. The previous page cursor is affected because a calculated new search window and current earliest departure determine the previous page cursor's earliest departure time. The next page check does not show in the diff here.

https://github.com/opentripplanner/OpenTripPlanner/blob/bb6bc6538239c55b26be149c41a1db92d7b12f75/application/src/main/java/org/opentripplanner/model/plan/paging/cursor/PageCursorFactory.java#L154-L156

optionsome commented on application/src/test/java/org/opentripplanner/service/paging/PS3_FewItinerariesOnSearchWindowLimitTest.java:
```diff
@@ -65,16 +72,20 @@ class PS3_FewItinerariesOnSearchWindowLimitTest {
   static List<Arguments> testCases() {
     return List.of(
       // Itineraries depart inside search-window, step NEXT 6 times
-      //Arguments.of("08:00", DEPART_AFTER, "1 N 2 N - N - N 3 N - N"),
+      Arguments.of("08:00", DEPART_AFTER, "1 N 2 N - N - N 3 N - N"),
       // Itineraries depart inside search-window, step PREV 6 times
-      //Arguments.of("08:00+1d", DEPART_AFTER, "3 P - P - P 2 P 1 P - P"),
+      Arguments.of("08:00+1d", DEPART_AFTER, "3 P - P - P 2 P 1 P - P"),
       // Itineraries depart inside search-window, step BACK and FORTH
-      //Arguments.of("08:00", DEPART_AFTER, "1 N 2 N - P 2 N - P 2 P 1 N 2 N"),
+      Arguments.of("08:00", DEPART_AFTER, "1 N 2 N - P 2 N - P 2 P 1 N 2 N"),
```
Why were these arguments commented out previously? I've trouble understanding what's going on in this test class in general with the pattern matching etc. Perhaps at least add a comment on the pattern for what it is used for. It's a bit difficult to read the regex and understand its usage in this context.

optionsome commented on application/src/test/java/org/opentripplanner/service/paging/TestPagingUtils.java:
```diff
@@ -13,7 +13,8 @@ static String cleanStr(Object value) {
       .replaceAll("2020-02-03T(\\d\\d:\\d\\d):00Z", "$1+1d")
-      .replaceAll("0(\\d:00)", "$1");
+      .replaceAll("0(\\d:00)", "$1")
+      .replaceAll("0(\\d:01)", "$1");
```
This class could also use some comments javadoc and maybe even comments to explain what each of these replaces do.

VillePihlava commented on application/src/main/java/org/opentripplanner/model/plan/paging/cursor/PageCursorFactory.java:
```diff
@@ -156,6 +171,9 @@ private Instant edtBeforeNewSw() {
 
   private Instant edtAfterUsedSw() {
+    if (firstSearchLatestItineraryDeparture != null) {
+      return firstSearchLatestItineraryDeparture.plusSeconds(60);
```
Lets talk about this in the next dev meeting

t2gran replied:
Raptor iterations are in 60s increments - unsure what happens here if you add only one second - it migh be rounded up or down - do not know.

t2gran commented on application/src/test/java/org/opentripplanner/service/paging/PS3_FewItinerariesOnSearchWindowLimitTest.java:
```diff
@@ -48,33 +51,47 @@ class PS3_FewItinerariesOnSearchWindowLimitTest {
+   *     - "08:00"            : The first search departure time
+   *     - DEPART_AFTER       : If first search is arriveBy or depart after search
+   *     - "1 N3600 2 N - .." : Paging sequence - N:NEXT or P:PREVIOUS, a number directly after
+   *                            N or P indicates a time shift value in seconds that differs from
```
```suggestion
   *                            N or P indicates a time shift value in seconds that differs from
```
Can you make this minutes - the search-window and Raptor iterations uses minutes, it does not have to, but lets deal with it if we decide that the minute iteration step is to large.

VillePihlava commented:
I added a constant for the time that is added to the exclusive end of the search window. I also added a check for making sure the `currentEdt` is not set past the search window.

About seconds being included in the `Instant` fields of the `PageCursor`:

https://github.com/opentripplanner/OpenTripPlanner/blob/9a0fc4d189742b492d1fcf4abbed5de94277dde6/application/src/main/java/org/opentripplanner/model/plan/paging/cursor/PageCursorFactory.java#L125-L128

Because seconds can also be included in the `prevEdt` here (e.g. `12:30:29`), I think adding the check mentioned above is enough and potentially adding seconds to the times shouldn't matter.

VillePihlava commented on application/src/main/java/org/opentripplanner/model/plan/paging/cursor/PageCursorConstants.java:
```diff
@@ -0,0 +1,13 @@
+   * <p>
+   * The value is set to 60 seconds because raptor operates in one minute increments.
+   */
+  public static final int SEARCH_WINDOW_END_EXCLUSIVITY_TIME_ADDITION_SECONDS = 60;
```
I changed the variable to a `Duration` and removed the `_SECONDS` from the name

t2gran commented:
Only one small comment above and I am ready to approve ✅.

t2gran commented on application/src/main/java/org/opentripplanner/service/paging/PagingService.java:
```diff
@@ -184,12 +186,33 @@ private PageCursorFactory mapIntoPageCursorFactory(@Nullable PageType currentPag
+   * If the first search is an arrive by search, add the latest itinerary's start time
+   * to allow setting cursor information correctly for the page cursor of the next page.
+   */
+  private void handleFirstSearch(@Nullable PageType currentPageType, PageCursorFactory factory) {
```
Consider: I think this method is an abstraction leak. All business logic to calculate the `PageCursor` should be in the `PageCursorFactory` the Service should not know about the details. To fix this, all you need to do is:
 -  make the `resolvePageTypeForTheFirstSearch` private
 - move this logic into `PageCursorFactory`
 - pass in `lastItineraryDeparture` to the factory. 
 
 Sorry for not pointing out this before - I only had a hunch there was something wrong here, and only after a break did it appear to me what was wrong 🤔.

VillePihlava replied:
I moved the logic into `PageCursorFactory`, I passed the itinerary list as a parameter to the `withOriginalSearch` method instead of `lastItineraryDeparture` because this seemed more logical.

VillePihlava replied:
I changed the `PageCursorFactory` to take an instant instead of the list of itineraries as we discussed in the dev meeting

t2gran reviewed (changes requested):
One other thing...

VillePihlava commented on application/src/main/java/org/opentripplanner/model/plan/paging/cursor/PageCursorConstants.java:
```diff
@@ -0,0 +1,15 @@
+   * <p>
+   * The value is set to 60 seconds because raptor operates in one minute increments.
+   */
+  public static final Duration SEARCH_WINDOW_END_EXCLUSIVITY_TIME_ADDITION = Duration.ofSeconds(60);
```
I moved the variable into PageCursorFactory

optionsome commented on application/src/main/java/org/opentripplanner/model/plan/paging/cursor/PageCursorFactory.java:
```diff
@@ -22,28 +22,47 @@ public class PageCursorFactory {
+   * <p>
+   * The value is set to 60 seconds because raptor operates in one minute increments.
+   */
+  private static final Duration SEARCH_WINDOW_END_EXCLUSIVITY_TIME_ADDITION = Duration.ofSeconds(
```
Static final variables should be at the top of the class

VillePihlava replied:
I moved the variable to the top and changed the wording from seconds -> minute
